pkg/auth/providers/azure/azure_provider.go

```diff
@@ -211,154 +217,154 @@ func (ap *azureProvider) TransformToAuthProvider(
+	for _, u := range users.Values() {
+		p, err := userToPrincipal(u)
+		if err != nil {
+			return nil, err
+		}
+		p.Me = samePrincipal(&token.UserPrincipal, p)
+		principals = append(principals, *p)
 	}
-
-	p := ap.groupToPrincipal(group)
-	p.MemberOf = ap.tokenMGR.IsMemberOf(token, p)
-
-	return p, nil
+	return principals, nil
 }
 
-func (ap *azureProvider) searchUsers(client *azureClient, name string, token v3.Token) ([]v3.Principal, error) {
-	filter := fmt.Sprintf("startswith(userPrincipalName,'%[1]s') or startswith(displayName,'%[1]s') or startswith(givenName,'%[1]s') or startswith(surname,'%[1]s')", name)
-	users, err := client.userClient.List(context.Background(), filter, "")
-	if err != nil {
-		return nil, err
+func (ap *azureProvider) msGraphGroupsToPrincipals(groups *[]msgraph.Group, token v3.Token) ([]v3.Principal, error) {
+	var principals []v3.Principal
+	for _, group := range *groups {
+		p, err := groupToPrincipal(&group)
+		if err != nil {
+			return nil, err
+		}
+		p.MemberOf = ap.tokenMGR.IsMemberOf(token, *p)
+		principals = append(principals, *p)
 	}
+	return principals, nil
+}
 
+func (ap *azureProvider) azureADGraphGroupsToPrincipals(groups graphrbac.GroupListResultPage, token v3.Token) ([]v3.Principal, error) {
 	var principals []v3.Principal
-	for _, user := range users.Values() {
-		p := ap.userToPrincipal(user)
-		p.Me = samePrincipal(token.UserPrincipal, p)
-		principals = append(principals, p)
+	for _, group := range groups.Values() {
+		p, err := groupToPrincipal(group)
+		if err != nil {
+			return nil, err
+		}
+		p.MemberOf = ap.tokenMGR.IsMemberOf(token, *p)
+		principals = append(principals, *p)
 	}
 	return principals, nil
 }
 
-func (ap *azureProvider) searchGroups(client *azureClient, name string, token v3.Token) ([]v3.Principal, error) {
+func (ap *azureProvider) searchGroups(client clients.AzureClient, name string, token v3.Token) ([]v3.Principal, error) {
 	filter := fmt.Sprintf("startswith(displayName,'%[1]s') or startswith(mail,'%[1]s') or startswith(mailNickname,'%[1]s')", name)
-	groups, err := client.groupClient.List(context.Background(), filter)
+	groups, err := client.ListGroups(filter)
 	if err != nil {
 		return nil, err
 	}
 
-	var principals []v3.Principal
-	for _, group := range groups.Values() {
-		p := ap.groupToPrincipal(group)
-		p.MemberOf = ap.tokenMGR.IsMemberOf(token, p)
-		principals = append(principals, p)
+	switch g := groups.(type) {
+	case *[]msgraph.Group:
+		return ap.msGraphGroupsToPrincipals(g, token)
+	case graphrbac.GroupListResultPage:
+		return ap.azureADGraphGroupsToPrincipals(g, token)
+	default:
+		return nil, errors.New("wrong type of group result")
 	}
-	return principals, nil
 }
 
-func (ap *azureProvider) newAzureClient(secret string) (*azureClient, error) {
-	config, err := ap.getAzureConfigK8s()
-	if err != nil {
-		return nil, err
+func (ap *azureProvider) newClient(cfg *v32.AzureADConfig, secret string) (clients.AzureClient, error) {
```

Comment:
You may want to consider making a wrapper around this function - I saw the pattern of "if old get user token, pass in secret" at least twice. Maybe something like:

```
func (ap *azureProvider) newClient(cfg *v32.AzureAdConfig, token v3.Token) (*clients.AzureClient, error){
        var secret string
	if authProviderOld(cfg) {
		secret, err = ap.tokenMGR.GetSecret(token.UserID, Name, []*v3.Token{&token})
		if err != nil && !apierrors.IsNotFound(err) {
		        return ap.newAzureADGraphClient(cfg, secret)
		}
	}

	return ap.newMSGraphClient(cfg)
```